# Importing relevant modules

In [1]:
from BirdSpeciesClassification import isBird
from os import listdir
from PIL import Image, ImageFilter, ImageDraw, ImageEnhance
from numpy.random import randint

# Bird Classification

## The 95% accuracy test

### on all images

In [2]:
# num_files,num_birds=0,0
# for directory in listdir('./data/'):
#     for file in listdir('data/' + directory):
#         num_files += 1
#         if isBird('data/' + directory + '/' + file):
#             num_birds += 1

instead of running it on all data, we decided to run it on 150 Wren images, 50 Leaf images, 50 House images, and 50 Grass images.

### on Wren images

In [10]:
num_files,num_birds=0,0
for (index,file) in enumerate(listdir('data/Wren/'),0):
    if index % 10 == 0:
        print(f"{index} image were classified")
    num_files += 1
    if isBird('data/Wren/' + file)[0]: # We have a 95% confidence by default.
        num_birds += 1
print(num_files,num_birds)

0 image were classified
10 image were classified
20 image were classified
30 image were classified
40 image were classified
50 image were classified
60 image were classified
70 image were classified
80 image were classified
90 image were classified
100 image were classified
110 image were classified
120 image were classified
130 image were classified
140 image were classified
150 136


We get 150 136 so 136/150 Birds were spotted in the images. 

That is almost a 91% test accuracy on this specific set.

### on Leaf images

In [11]:
num_files,num_birds=0,0
print("Leaf images")
for (index, file) in enumerate(listdir('data/Leaf/'),0):
    if index < 50:
        if index % 10 == 0:
            print(f"{index} image were classified")
        num_files += 1
        if isBird('data/Leaf/' + file)[0]: # We have a 95% confidence by default.
            num_birds += 1
print(num_files,num_birds)

Leaf images
0 image were classified


10 image were classified
20 image were classified
30 image were classified
40 image were classified
50 0


We get 50 0 so 0/0 Birds were spotted in the images. 

That is a 100% test accuracy.

### on House images

In [12]:
num_files,num_birds=0,0
print("House images")
for (index, file) in enumerate(listdir('data/House/'),0):
    if index < 50:
        if index % 10 == 0:
            print(f"{index} image were classified")
        num_files += 1
        if isBird('data/House/' + file)[0]: # We have a 95% confidence by default.
            num_birds += 1
print(num_files,num_birds)

House images
0 image were classified
10 image were classified
20 image were classified
30 image were classified
40 image were classified
50 0


We get 50 0 so 0/0 Birds were spotted in the images. 

That is a 100% test accuracy.

### on Grass images

In [13]:
num_files,num_birds=0,0
print("Grass images")
for (index, file) in enumerate(listdir('data/Grass/'),0):
    if index < 50:
        if index % 10 == 0:
            print(f"{index} image were classified")
        num_files += 1
        if isBird('data/Grass/' + file)[0]: # We have a 95% confidence by default.
            num_birds += 1
print(num_files,num_birds)

Grass images
0 image were classified
10 image were classified
20 image were classified
30 image were classified
40 image were classified
50 0


We get 50 0 so 0/0 Birds were spotted in the images. 

That is a 100% test accuracy.

## Conclusion

All in all, that is a 95% balanced accuracy on the test set with 150 bird images and 150 non-bird images. 

Also, the Grass images take longer because they are relatively large inputs to the machine learning model.

# Bird Species Classification

## The 50% accuracy tests

In [20]:
num_files,num_birds=0,0
print("Feral Pigeon images")
for (index, file) in enumerate(listdir('data/FeralPigeon/'),0):
    if index < 20:
        if index % 10 == 0:
            print(f"{index} image were classified")
        num_files += 1
        bird, species = isBird('data/FeralPigeon/' + file)
        if species is not None: # We have a 95% confidence by default.
            num_birds += 1
        print(species)
print(num_files,num_birds)

Feral Pigeon images
0 image were classified
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
10 image were classified
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
20 20


We get 20 Pigeon as well as 20 20 so 20/20 Birds were spotted in the images. 

Since all images contained a Pigeon, we have a 100% test accuracy on this specific test set.

## Extending the bird species detection test to multiple bird species

In [21]:
num_files,num_birds,num_species=0,0,0
for directory in listdir('./data/'):
    if directory not in ["Grass","Leaf","House"]:
        print(f"We enter directory {directory}")
        for (index,file) in enumerate(listdir('data/' + directory),0):
            if index < 5:
                num_files += 1
                bird, species = isBird('data/' + directory + '/' + file)
                if bird:
                    num_birds += 1
                    if species is not None:
                        num_species += 1
                        print(species)
print(f"We looked at {num_files} files")
print(f"We found {num_birds} birds")
print(f"We found {num_species} species")

We enter directory Dunnock
Finch
Finch
Finch
Finch
Finch
We enter directory WoodPigeon
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
We enter directory Goldfinch
Finch
Finch
Finch
Finch
Finch
We enter directory Wren
Wren
Wren
Wren
Wren
Wren
We enter directory Robin
Robin
Robin
Robin
Robin
Robin
We enter directory HouseSparrow
Sparrow
Sparrow
Sparrow
Sparrow
Sparrow
We enter directory Bluetit
Finch
Finch
Finch
Finch
Finch
We enter directory FeralPigeon
Pigeon
Pigeon
Pigeon
Pigeon
Pigeon
We enter directory SongThrush
Blackbird
Blackbird
Blackbird
We enter directory Greenfinch
Finch
Finch
Finch
Finch
Finch
We enter directory Magpie
Magpie
Magpie
Magpie
Magpie
Magpie
We enter directory CoalTit
Finch
Finch
Finch
Finch
Finch
We enter directory Jackdaw
Blackbird
Blackbird
Blackbird
Crow
Blackbird
We enter directory CollaredDove
Pigeon
Pigeon
Pigeon
Pigeon
We enter directory Blackbird
Blackbird
Blackbird
Blackbird
Blackbird
Blackbird
We enter directory CarrionCrow
Blackbird
Blackbird
Crow
Blackbird
Black

We provide 100 images that contain a bird and the model identifies that there ar 99 birds in the images with 95% confidence. Additionally, the model has detected 95 bird species that are also in the bird_species list as defined for our project. Not all of these species are correct. Manually, I found out that 10 out of 20 species were perfectly detected. Unfortunately, we cannot classify the Dunnock, Bluetit, SongThrush, CoalTit, Jackdaw, LongTailedTit, GreatTit, and Starling. Since the ML model has 12 labels that fit these common UK (and NL) birds.

## Conclusion

We find We get a 60 percent accuracy on the test set of 100 images for 20 classes.
Since the ML model has 12 labels that fit these common UK (and NL) birds. Unfortunately, we cannot classify the Dunnock, Bluetit, SongThrush, CoalTit, Jackdaw, LongTailedTit, GreatTit, and Starling.

# Input Size

We want to identify whether a different size of the input image will have a significant effect on the accuracy of the bird classication task. We have decided to test this by regarding the output of isBird before and after resizing an image.

In [41]:
confusion_matrix = [[0,0],[0,0]]
for directory in listdir('./data/'):
    if directory not in ["Grass","Leaf","House"]:
        print(f"We enter directory {directory}")
        for (index,file) in enumerate(listdir('data/' + directory),0):
            if index > 9 and index < 15:
                real = isBird('./data/' + directory + '/' + file)[0]
                print(real)
                IMG = Image.open('data/' + directory + '/' + file)
                IMG = IMG.resize((512,512))
                IMG.save('data/' + directory + '/' + file)
                pred = isBird('data/' + directory + '/' + file)[0]
                print(pred)
                confusion_matrix[real][pred] = confusion_matrix[real][pred] + 1
print(confusion_matrix)

We enter directory Dunnock
True
True
True
True
True
True
True
True
True
True
We enter directory WoodPigeon
True
True
True
True
True
True
True
True
True
True
We enter directory Goldfinch
True
True
True
True
True
True
True
True
True
True
We enter directory Wren
True
True
True
True
True
True
True
True
True
True
We enter directory Robin
True
True
True
True
True
True
True
True
True
True
We enter directory HouseSparrow
True
True
True
True
True
True
True
True
True
True
We enter directory Bluetit
True
True
True
True
True
True
True
True
True
True
We enter directory FeralPigeon
True
True
True
True
True
True
True
True
True
True
We enter directory SongThrush
True
True
True
True
True
True
True
True
True
True
We enter directory Greenfinch
True
True
True
True
True
True
True
True
True
True
We enter directory Magpie
True
True
True
True
True
True
True
True
True
True
We enter directory CoalTit
True
True
True
True
True
True
True
True
True
True
We enter directory Jackdaw
True
True
True
True
True
True
True


In the 2D confusion matrix, we see that all results are True Positives, so all images were correctly classified both before and after resizing.

This shows that the model can handle various sizes of the input image. So we do not need to resize images. 

On the other hand, we could resize the images to speed up the isBird function because posting larger images takes more time.

# Blurry condensation

We want to identify whether blurring the input image will have a significant effect on the accuracy of the bird classication task. We have decided to test this by regarding the output of isBird before and after blurring an image.

In [36]:
confusion_matrix = [[0,0],[0,0]]
for directory in listdir('./data/'):
    if directory not in ["Grass","Leaf","House"]:
        print(f"We enter directory {directory}")
        for (index,file) in enumerate(listdir('data/' + directory),0):
            if index < 5:
                real = isBird('./data/' + directory + '/' + file)[0]
                print(real)
                IMG = Image.open('data/' + directory + '/' + file)
                IMG = IMG.filter(ImageFilter.GaussianBlur(radius=5))
                IMG.save('data/' + directory + '/' + file)
                pred = isBird('data/' + directory + '/' + file)[0]
                print(pred)
                confusion_matrix[real][pred] = confusion_matrix[real][pred] + 1
print(confusion_matrix)

We enter directory Dunnock
True
True
True
True
True
True
True
True
True
True
We enter directory WoodPigeon
True
False
True
True
True
True
True
True
True
True
We enter directory Goldfinch
True
True
True
True
True
True
True
True
True
True
We enter directory Wren
True
True
True
True
True
True
True
True
True
True
We enter directory Robin
True
True
True
True
True
True
True
True
True
True
We enter directory HouseSparrow
True
True
True
True
True
True
True
True
True
True
We enter directory Bluetit
True
True
True
True
True
True
True
True
True
True
We enter directory FeralPigeon
True
True
True
True
True
True
True
True
True
True
We enter directory SongThrush
True
True
True
True
True
True
True
True
True
True
We enter directory Greenfinch
True
True
True
True
True
True
True
True
True
True
We enter directory Magpie
True
True
True
True
True
True
True
True
True
True
We enter directory CoalTit
True
True
True
True
True
True
True
True
True
True
We enter directory Jackdaw
True
True
True
True
True
True
True

In the 2D confusion matrix, there is a large value in the right bottom entry and low values in the other entries. Out of 100 images, 98 images were classified correctly before and after blurring. We misclassified one image of a Collared Dove twice (before and after blurring). We missclassified one image of a WoodPigeon (after blurring).

This shows that the model is robust to blurring. So the model will do well at detecting birds in images even when there is condensation.

# Artificial rain

We want to identify whether adding some noise in the form of rain-like patterns to the input image will have a significant effect on the accuracy of the bird classication task. We have decided to test this by regarding the output of isBird before and after adding this noise to an image.

In [39]:
def generate_random_lines(imshape,slant,drop_length):
  drops=[]
  for _ in range(1500): ## If You want heavy rain, try increasing this
    if slant<0:
      x = randint(slant,imshape[1])
    else:
      x = randint(0,imshape[1]-slant)
    y = randint(0,imshape[0]-drop_length)
    drops.append((x,y))
  return drops

def add_rain(image):
  imshape = image.size
  slant_extreme = 10
  slant= randint(-slant_extreme,slant_extreme)
  drop_length=20
  drop_width=2
  drop_color=(200,200,200) ## a shade of gray
  rain_drops= generate_random_lines(imshape,slant,drop_length)
  for rain_drop in rain_drops:
    draw = ImageDraw.Draw(image)
    draw.line((rain_drop[0],rain_drop[1], rain_drop[0]+slant,rain_drop[1]+drop_length), fill=drop_color, width=drop_width)
  image = image.filter(ImageFilter.GaussianBlur(radius=1))
  filter = ImageEnhance.Brightness(image)
  image = filter.enhance(0.70)
  return image

In [40]:
confusion_matrix = [[0,0],[0,0]]
for directory in listdir('./data/'):
    if directory not in ["Grass","Leaf","House"]:
        print(f"We enter directory {directory}")
        for (index,file) in enumerate(listdir('data/' + directory),0):
            if index > 4 and index < 10:
                real = isBird('./data/' + directory + '/' + file)[0]
                print(real)
                IMG = Image.open('data/' + directory + '/' + file)
                IMG = add_rain(IMG)
                IMG.save('data/' + directory + '/' + file)
                pred = isBird('data/' + directory + '/' + file)[0]
                print(pred)
                confusion_matrix[real][pred] = confusion_matrix[real][pred] + 1
print(confusion_matrix)

We enter directory Dunnock
True
True
True
True
True
True
True
True
True
True
We enter directory WoodPigeon
True
True
True
True
True
False
True
False
True
False
We enter directory Goldfinch
True
True
True
True
True
True
True
True
True
True
We enter directory Wren
True
True
True
True
True
True
True
True
True
True
We enter directory Robin
True
False
True
True
True
False
True
True
True
False
We enter directory HouseSparrow
True
True
True
False
True
True
True
True
True
True
We enter directory Bluetit
True
True
True
True
True
True
True
True
True
False
We enter directory FeralPigeon
True
True
True
False
True
True
True
False
True
True
We enter directory SongThrush
True
True
True
True
False
True
True
True
True
True
We enter directory Greenfinch
True
True
True
True
True
True
True
True
True
True
We enter directory Magpie
True
True
True
False
True
False
True
True
True
True
We enter directory CoalTit
True
False
True
False
True
False
True
False
True
False
We enter directory Jackdaw
True
False
True
F

In the 2D confusion matrix, there is a large value in the right bottom entry and low values in the other entries. Out of 100 images, 70 images were classified correctly before and after blurring. We misclassified 30 images and most of those misclassifications happened because the addition of rain effects made it more difficult to detect a bird in the image. 

This shows that the model is quite robust to slight blurring, brightness lowering, and the addition of rain-like grey lines. So the model will do well at detecting birds in images even when the outdoor environment is rainy.

## Conclusion

The machine learning model is robust and can handle different sizes of input images as well as the effects of various weather conditions to the images.

In [12]:
import requests
import datetime
import json

ip = '145.126.60.24'
port = '8080'

auth = 'eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJyYXNwaSIsImFjY291bnRfaWQiOjcyLCJleHAiOjE2OTk0NTYyNDJ9.YVjfzVs4Y9QyVdqUXTw-D25CpIlCtYNemHs7ALz-xP8'

timestamp = str(datetime.datetime.now())

# timestamp = datetime.datetime.fromtimestamp(timestamp)
# timestamp = datetime.datetime.strftime('%Y-%m-%d %H:%M:%S')

visit_id = 69
filepath = 'data/HouseSparrow/2.jpg'
bird,species,confidence = isBird(filepath)

base_url = f'http://{ip}:{port}/bad/api'

if not bird or not species:
    # add visit with images
    r = requests.put(base_url + '/images/archive/visits',
                     json = {
                        'visitId':visit_id,
                        "species":"No Bird",
                        "arrival":timestamp,
                        "departure":timestamp,
                        "visitLen":310,
                        "accuracy":95.12,
                        "images":[ 
                            {
                                "imageId": 1,
                                "visitId": 1,
                                "date": timestamp,
                                "image_path": "meta/img/main/bird1.jpeg"
                            }
                        ]
                     },
                     headers = {
                        'Authorization':auth,
                        'Content-Type':'application/json'
                        }
                    )

else:
    # add visit with images
    r = requests.put(base_url + '/images/main/visits',
                     json = {
                        'visitId':visit_id,
                        "species":species,
                        "arrival":timestamp,
                        "departure":timestamp,
                        "visitLen":310,
                        "accuracy":confidence,
                        "images":[ 
                            {
                                "imageId": 1,
                                "visitId": 1,
                                "date": timestamp,
                                "image_path": "meta/img/main/bird1.jpeg"                            }
                        ]
                     },
                     headers = {
                        'Authorization':auth,
                        'Content-Type':'application/json'
                        }
                    )

if r != 200:
    # add filepath to backlog
    filepath
    print(r)
    print(r.status_code)
else:
    print("Yay!")

TypeError: Object of type datetime is not JSON serializable